orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -869,89 +869,217 @@ common::Status HandleSharedInitializer(Graph& graph,
+  // A tensor produced in stage r and consumed in stage r', such that r' > r,
+  // will have a replica in all stages r'', such that r'' > r and r'' < r'.
+  // tensor_replicas[t][r] contains a pointer to the the replica of t in stage r, 
+  // it if exists, or to itself if r is the stage of the producer of r.
+  std::map<const NodeArg*, std::vector<NodeArg*>> tensor_replicas;
+  auto create_tensor_replica = [&tensor_replicas, &graph](const NodeArg* tensor,
+                                                          int consumer_stage) {
+    NodeArg& new_receive_output = CreateNodeArg(graph, *tensor);
+    const auto* old_shape = tensor->Shape();
+    if (old_shape != nullptr) {
+      new_receive_output.SetShape(*old_shape);
+    }
+    // Add value info for this newly added receive_output, for shape propagation
+    // when training this partition.
+    graph.AddValueInfo(&new_receive_output);
+    tensor_replicas[tensor][consumer_stage] = &new_receive_output;
+  };
+
+  // Checks whether the tensor is produced and consumed in the forward stage of
+  // the computation.
+  auto is_forward = [](const int producer_stage, const int consumer_stage) {
+    return producer_stage < consumer_stage;
+  };
+
+  // Checks whether the tensor is produced and consumed in the backward stage of
+  // the computation. 
+  auto is_backward = [](const int producer_stage, const int consumer_stage) {
+    return producer_stage > consumer_stage;
+  };
+
+  // Find tensors that need to be sent and forwarded.
+  for (const NodeArg* node_arg : initial_node_args) {
+    // Initialize tensor_replicas data structure.
+    auto inserted = tensor_replicas.emplace(
+                      std::make_pair(node_arg, std::vector<NodeArg*>(nstages)));
+    auto& replicas = (*inserted.first).second;
+
+    // TODO: for now we pretend that inputs are produced in stage 0,
+    // but I need to double check how they are handled.
+    int producer_stage = 0;
+    Node* producer_node = graph.GetMutableProducerNode(node_arg->Name());
+    assert(producer_node != nullptr);
+    producer_stage = op_to_stage.find(producer_node)->second;
+    
+    auto consumers = graph.GetMutableConsumerNodes(node_arg->Name());
+    if (consumers.size() == 0) {
+      continue;
+    }
+    
+    // This is only handling forwarding in the forward part of the graph.
+    int last_consumer_stage_fwd = -1;
+    for (Node* consumer : consumers) {
+      auto found_stage = op_to_stage.find(consumer);
+      assert(found_stage != op_to_stage.end());
+      int consumer_stage = found_stage->second;
+      // TODO: test case in which a tensor is produced by a fwd op, stashed and
+      // sent to the previous stage by a bwd op.
+      // For now, I'm assuming that if a tensor is produced by a fwd op and
+      // consumed by a bwd op, then producer and consumer are both in the same
+      // device. This will not always be the case though. 
+      if (!IsBackward(*producer_node) && IsBackward(*consumer)) {
```

Comment:
>IsBackward(*consumer) [](start = 41, length = 21)

Is this function for both forward and forward+backward partitions?

Reply:
For now, this function works only for the forward graph. However, I believe that the loss ops will appear as backward operators. 

The error [function](https://github.com/microsoft/onnxruntime/blob/master/orttraining/orttraining/models/mnist/main.cc#L191) seems to be called only on the last device and if we don't make sure that the fetched tensors are in that device, we will get an error, at least with this implementation. 

Imagine the case in which [predictions](https://github.com/microsoft/onnxruntime/blob/master/orttraining/orttraining/models/mnist/main.cc#L197) is produced in device 3, but the error function runs in the device 4 (last device). For this to work, we could change the fetched tensors vector to point to the received replica of predictions. But for simplicity, we just require that predictions will be in the last device too.